### PR TITLE
feat: Add reasoning budget support to Bedrock models for extended thinking (#4201)

### LINKED ISSUE #4201: AWS Bedrock does not show or support Sonnet and Opus 4 Extended Thinking
### App Version

v3.19.1

### API Provider

AWS Bedrock

### Model Used

Claude Sonnet 4

### 🔁 Steps to Reproduce

Configration shown here: List does not showing :thinking variant

Image

Image

### 💥 Outcome Summary

List of models does not show :thinking variants for Anthropic Models. This is contrary to what was stated in https://github.com/RooCodeInc/Roo-Code/pull/3844

Typing in various prompts after selecting Sonnet 4 also never trigger reasoning/extended thinking.

Model should either show :thinking variant or should have Enable Reasoning Mode (Similar to Anthropic as Provider) with Max Thinking Tokens configuration

Image

### 📄 Relevant Logs or Errors (Optional)

```shell

```

### PR DESCRIPTION
**Pull request by hannesrudolph**
## Description

Fixes #4201

Adds comprehensive Extended Thinking (reasoning budget) support for AWS Bedrock Claude models (Sonnet 4, Opus 4, and 3.7 Sonnet). This implementation provides the equivalent functionality to the "Enable Reasoning Mode" available in the direct Anthropic provider, but integrated with AWS Bedrock's API structure.

### 🎯 Core Implementation
- **Added `supportsReasoningBudget: true`** to Claude models in [`packages/types/src/providers/bedrock.ts`](packages/types/src/providers/bedrock.ts:76)
- **Implemented extended thinking support** in [`src/api/providers/bedrock.ts`](src/api/providers/bedrock.ts:315) with:
  - `additionalModelRequestFields.thinking` parameter support
  - Proper `anthropic_version` header handling
  - Dynamic budget token configuration
  - Conditional parameter exclusion (topP) when thinking is enabled

### 🔄 Stream Processing Enhancements
- **Enhanced stream event handling** to properly process reasoning content:
  - Support for `reasoningContent` blocks
  - Support for `thinking` content types
  - Proper content block indexing and formatting
  - Multiple fallback structures for AWS SDK compatibility

### 🧪 Comprehensive Testing
- **Added extensive test suite** in [`src/api/providers/__tests__/bedrock-reasoning.test.ts`](src/api/providers/__tests__/bedrock-reasoning.test.ts) with 280 lines of tests covering:
  - Extended thinking parameter inclusion for supported models
  - Parameter exclusion for non-supported models
  - Stream processing for reasoning content
  - Multiple event structure compatibility
  - Error handling and validation

### 🎨 UI Improvements
- **Updated ThinkingBudget component** in [`webview-ui/src/components/settings/ThinkingBudget.tsx`](webview-ui/src/components/settings/ThinkingBudget.tsx:69) to:
  - Properly calculate max token limits for Bedrock models
  - Show reasoning budget controls when `supportsReasoningBudget` is true
  - Integrate with existing reasoning effort settings
- **Fixed ApiOptions component** to conditionally render ThinkingBudget only when models are available
- **Improved Bedrock provider settings** layout and conditional rendering

### 🛠️ Error Handling
- **Added VALIDATION_ERROR handling** for parameter validation issues
- **Enhanced error messaging** with specific guidance for extended thinking problems
- **Improved logging** for debugging extended thinking requests

### API Integration
The implementation follows AWS Bedrock's official extended thinking specification:
```typescript
additionalModelRequestFields: {
  thinking: {
    type: "enabled",
    budget_tokens: metadata?.thinking?.maxThinkingTokens || modelConfig.reasoningBudget || 4096
  }
}
```

### Stream Processing
Handles multiple AWS SDK event structures:
- `contentBlock.reasoningContent.text` (official AWS docs structure)
- `contentBlock.type === "thinking"` (alternative structure)
- `delta.reasoningContent.text` and `delta.thinking` (delta events)

### Model Support
Enabled for:
- `anthropic.claude-sonnet-4-20250514-v1:0`
- `anthropic.claude-opus-4-20250514-v1:0` 
- `anthropic.claude-3-7-sonnet-20250219-v1:0`

## Testing Performed

- ✅ **All existing tests pass** - verified backward compatibility
- ✅ **New test suite passes** - 280 lines of reasoning-specific tests
- ✅ **Manual testing completed**:
  - Extended thinking activates correctly for supported models
  - Reasoning content renders properly in the UI
  - Settings UI shows/hides controls appropriately
  - Parameter validation works correctly
  - Error handling provides helpful messages

## Verification of Acceptance Criteria

From issue #4201:
- ✅ **Models show reasoning support**: Instead of :thinking variants, models now support reasoning budget configuration
- ✅ **Enable Reasoning Mode equivalent**: ThinkingBudget component provides the same functionality as Anthropic provider
- ✅ **Max Thinking Tokens configuration**: Fully configurable budget with appropriate limits
- ✅ **Extended thinking triggers**: Reasoning now activates correctly when enabled and supported
- ✅ **UI consistency**: Matches the pattern established by other providers

## Checklist

- ✅ Code follows project style guidelines
- ✅ Self-review completed
- ✅ Comments added for complex logic (stream processing, reasoning detection)
- ✅ No breaking changes to existing Bedrock functionality
- ✅ Comprehensive test coverage for new functionality
- ✅ Error handling covers edge cases and provides helpful messages
- ✅ UI changes are consistent with existing design patterns
- ✅ Performance impact minimal (only adds parameters when needed)

## Related Issues

- Fixes #4201: AWS Bedrock does not show or support Sonnet and Opus 4 Extended Thinking
- Related to #4279: Support extended thinking for Claude 3.7/4 models in Bedrock (closed as duplicate)

## Screenshots/Demo

**Before**: No reasoning budget controls visible for Bedrock models
**After**: ThinkingBudget component appears for supported Bedrock models, matching Anthropic provider functionality

Users can now:
1. Select a supported Claude model in Bedrock
2. See the "Enable Reasoning Mode" toggle and token budget slider
3. Configure reasoning parameters just like with the direct Anthropic provider
4. Experience extended thinking in their conversations with proper reasoning content rendering

----

> [!IMPORTANT]
> Adds reasoning budget support to AWS Bedrock models, enabling extended thinking capabilities with updated configurations, stream processing, UI, and error handling.
> 
>   - **Behavior**:
>     - Adds `supportsReasoningBudget: true` to Claude models in `bedrock.ts`.
>     - Implements extended thinking in `bedrock.ts` with `additionalModelRequestFields.thinking` and dynamic budget token configuration.
>     - Excludes `topP` parameter when thinking is enabled.
>   - **Stream Processing**:
>     - Enhances stream event handling in `bedrock.ts` to process `reasoningContent` and `thinking` content types.
>   - **Testing**:
>     - Adds tests in `bedrock-reasoning.test.ts` for reasoning parameter inclusion, stream processing, and error handling.
>   - **UI**:
>     - Updates `ThinkingBudget.tsx` to calculate max token limits and show reasoning budget controls.
>     - Updates `ApiOptions.tsx` to conditionally render `ThinkingBudget`.
>   - **Error Handling**:
>     - Adds `VALIDATION_ERROR` handling for parameter validation issues in `bedrock.ts`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=RooCodeInc%2FRoo-Code&utm_source=github&utm_medium=referral)<sup> for dafa25b4b0825f17023ab819efeae3f07d8ba93f. You can [customize](https://app.ellipsis.dev/RooCodeInc/settings/summaries) this summary. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: webview-ui/src/components/settings/ThinkingBudget.tsx [hunk 1]
```diff
@@ -65,7 +69,21 @@ export const ThinkingBudget = ({ apiConfiguration, setApiConfigurationField, mod
...
+												customMaxOutputTokens,
+												DEFAULT_HYBRID_REASONING_MODEL_MAX_TOKENS,
+											)
+										: apiConfiguration.apiModelId
+											? getModelMaxOutputTokens({
+													modelId: apiConfiguration.apiModelId,
+													model: modelInfo,
+													settings: apiConfiguration,
+												})
+											: modelInfo.maxTokens
```
Comment (daniel-lxs) [inline webview-ui/src/components/settings/ThinkingBudget.tsx]: Is this ternary logic intentional? Since the component only renders when `isReasoningBudgetSupported` is true (line 52), the `else` branches here (lines 79-85) would never execute. Could we simplify this to just use the first branch since we know reasoning budget is supported at this point?

File: src/api/providers/bedrock.ts [hunk 2]
```diff
@@ -280,18 +324,47 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
 			conversationId,
 		)
 
-		// Construct the payload
+		let additionalModelRequestFields: BedrockThinkingConfig | undefined
+		let thinkingEnabled = false
+
+		if (
```
Comment (daniel-lxs) [inline src/api/providers/bedrock.ts]: This condition seems quite complex with multiple OR branches. Could we break this down for clarity? Also, is there a specific reason why we check both `metadata?.thinking?.enabled` AND the `shouldUseReasoningBudget` logic? When would one be true but not the other?

File: src/api/providers/bedrock.ts [hunk 3]
```diff
@@ -397,19 +470,69 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
 				}
 
 				// Handle content blocks
-				if (streamEvent.contentBlockStart?.start?.text) {
-					yield {
-						type: "text",
-						text: streamEvent.contentBlockStart.start.text,
+				if (streamEvent.contentBlockStart) {
+					// The actual AWS SDK structure based on the documentation
+					const cbStart = streamEvent.contentBlockStart as any
```
Comment (daniel-lxs) [inline src/api/providers/bedrock.ts]: Using `as any` here bypasses TypeScript safety. Could we define proper types for the AWS SDK stream events instead? This would help catch potential structure changes and make the code more maintainable.

File: src/api/providers/bedrock.ts [hunk 4]
```diff
@@ -397,19 +470,69 @@ export class AwsBedrockHandler extends BaseProvider implements SingleCompletionH
...
+						if (cbStart.contentBlockIndex && cbStart.contentBlockIndex > 0) {
+							yield { type: "reasoning", text: "\n" }
+						}
+						yield {
+							type: "reasoning",
+							text: cbStart.contentBlock.reasoningContent.text || "",
+						}
+					}
+					// Check for thinking block - handle both possible structures
+					else if (cbStart.contentBlock?.type === "thinking" || cbStart.content_block?.type === "thinking") {
```
Comment (daniel-lxs) [inline src/api/providers/bedrock.ts:498]: The stream processing handles multiple possible structures (`contentBlock` vs `content_block`, `reasoningContent` vs `thinking`). Is this defensive programming for different AWS SDK versions, or are these all valid current structures? Some documentation about which structure is expected vs fallback would be helpful.

**Review by daniel-lxs — APPROVED** [review]
LGTM with emphasis on "looks" cause I don't have a way to test this.

**Comment by hannesrudolph** [comment]
I tested it

**Review by mrubens — APPROVED** [review]
